### ISSUE TITLE
Controlling webview

### ISSUE THREAD

**Issue by sergio-doman** [eclipsesource/tabris-js#667]
Is there any way to control webview? Some kind of callback with passing parameters, in both directions: from tabris to web view and back. 
For example, I need to update some part of content inside webview by tabris.

**Comment by cookieguru** [comment]
Depends on how you need to control it.  You can change the WebView's URL and set HTML dynamically:
```js
var page = tabris.create("Page", {
	title: "Web View",
	topLevel: true
});

var webview = tabris.create("WebView", {
	layoutData: {left: 0, top: 0, right: 0, bottom: 0}
}).appendTo(page);

webview.set("url", 'https://tabrisjs.com/');
webview.on('load', function(_webview) {
	console.warn(_webview.get('url')); //fires every page navigation
});
//do something
webview.set("html", '<b>Hello World!</b>');
```
The WebView acts just like an iframe in that it can control itself (navigate to different URLs) but cannot update anything outside of itself.

If you need to load local content and can't set the HTML dynamically, looks like you'll have to wait for #502.  #638 might also be of interest.

I just discovered the `load` event--it's undocumented and was introduced in #148 and will (for my purposes) close #575.

**Comment by PierBover** [comment]
Is it possible to send data between the JS inside the Webview and the JS of the main App?

**Comment by cookieguru** [comment]
@PierBover What kind of data?

**Comment by PierBover** [comment]
Objects, function calls, etc.

Obviously I mean from local HTML files.

**Comment by cookieguru** [comment]
@PierBover That's not normally something you can do with webpages period.

You can, however, set the _entire_ contents of the web view programmatically.  See the last line of the code block above.

**Comment by PierBover** [comment]
> That's not normally something you can do with webpages period.

Yes I understand that, which is why I'm asking.

> You can, however, set the entire contents of the web view programmatically

Yes I saw that on your previous comment, but what I need is to display some interactive activities made in HTML/JS inside a webview. Once the user has finished the activity send some data back to the main app and close the webview. Like say a mini game, and when the users finishes it we need to store his score.

Is this possible with Tabris.js?

**Comment by cookieguru** [comment]
You could try something like `webview.get('html');`

**Comment by sergio-doman** [comment]
Here I tried to describe situation, when you really need communication between app and webview  #598 
and you definitely can't solve the issue with `webview.get('html');`

**Comment by PierBover** [comment]
BTW in iOS `WKWebView` offers full fledged communication between JS in a WebView and native code.

In Android the class `WebView` also offers full fledged communication.

**Comment by tbuschto** [comment]
We will work on this. 

I suggest we go with the [HTML5 message API](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage). It's flexible and secure. Here is what it would look like:

Tabris.js
```
webView.postMessage(messageData, "*");
webView.on("message", function(event) {
  console.log("message from " + event.source + ":" + event.origin);
});
```

The website:
```
window.postMessage(messageData, "*");
window.addEventListener("message", function(event) {
  console.log("message from " + event.source + ":" + event.origin);
});
```

The standard demands a deep-cloning of the messageData. We could go with just JSON.stringify for the first version, which means things like Data and ArrayBuffer would not work yet.

Not sure the second parameter in postMessage needs to be supported, it basically just allows you to send the message to a specific domain (which we would have to invent for the webView->tabris.js direction.) "*" allows all domains.

Of course in addition an `eval`method could be implemented on webView. It's not nice to mess with other peoples websites, but since it's allowed on the native platforms we don't have any good reason to prevent it.

```
webView.eval("document.body.innerText='Boo!';");
```

**Comment by mpost** [comment]
@tbuschto If we were to hijack the postMessage mechanism wouldn't that render it useless for inter iframe communication inside the website?

**Comment by mpost** [comment]
After some discussion we agreed on the solution similar to the above with one adjustment: the website would use the parent (`window.parent.postMessage`) to interface with the native app:

Tabris.js
```js
webView.postMessage(messageData, "*");
webView.on("message", function(event) {
  console.log("message from " + event.source + ":" + event.origin);
});
```

The website:
```js
window.parent.postMessage(messageData, "*");
window.addEventListener("message", function(event) {
  console.log("message from " + event.source + ":" + event.origin);
});
```

**Comment by sergio-doman** [comment]
Wow, really good news, thanks. But in which Tabris version could be implemented this mechanism?

**Comment by mpost** [comment]
@sergio-doman We are targeting the upcoming Tabris.js 1.9 release.

**Comment by gosia-eclipsesource** [comment]
@sergio-doman @PierBover @cookieguru Web messaging is available in 1.9. There's a [snippet](https://github.com/eclipsesource/tabris-js/tree/master/snippets/webview-webmessaging) if you'd like to have a look.
